[fonts] allow different available fonts at the same time

AlwinEsch opened the pull request:
## Description

This adds a support to load two independent TTF fonts, background is to get a missing character from the second font.

It mainly refers to having supported skins and their fonts in all languages.

The following languages ​​cannot currently be displayed:
- Amharic
- Arabic (but works in Arial)
- Burmese
- Chinese (Simple) (but works in Arial)
- Chinese (Traditional) (but works in Arial)
- Hindi (Devanagiri)
- Japanese (but works in Arial)
- Kannada
- Korean (but works in Arial)
- Malayalam
- Persian (but works in Arial)
- Persian (Iran) (but works in Arial)
- Sinhala
- Tamil (India)
- Telugu
- Thai

The way is not the best, but it would be a solution.

The best solution would be to lay down a table in Kodi in which UTF8 character code series is assigned to a font and thus to select each individual character in the appropriate font (similar to Pango or how Chromium does it for websites).
However, Kodi's font system in the current style cannot be implemented in this way and would need a complete rebuild of the system (as far as I know).

The first commit renames a part, I was really confused how the name was used before because the same name (only with case difference) was used twice for different things.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context

To become some issues fixed. 

Not sure if this way acceptable, if not, then we can close request.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

![Bildschirmfoto vom 2021-11-19 21-23-39](https://user-images.githubusercontent.com/6879739/142745775-f6f87d3f-9a86-4079-b0f2-e8fcd53849a2.png)
![Bildschirmfoto vom 2021-11-20 00-24-38](https://user-images.githubusercontent.com/6879739/142745776-139496e9-0b2c-474e-96b5-2ca9afb27bdc.png)
![Bildschirmfoto vom 2021-11-21 02-28-30](https://user-images.githubusercontent.com/6879739/142745777-fb324a52-0ec1-4008-a8a0-df33ced99b64.png)
![Bildschirmfoto vom 2021-11-21 02-30-08](https://user-images.githubusercontent.com/6879739/142745799-d3a0570d-ae01-48c4-a80f-4f6bb7a3e8b0.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

enen92 commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -137,19 +157,22 @@ CGUIFont* GUIFontManager::LoadTTF(const std::string& strFontName,
       std::shared_ptr<CFontResource> font(std::static_pointer_cast<CFontResource>(it));
       if (font->GetFont(file, strPath))
+      {
+        strPathFallback = ""; // Reset fallback as we not know what addon brings
```
strPathFallback.clear()

AlwinEsch replied:
Thanks, is better and have changed!

AlwinEsch commented:
Related issues I found:
- https://github.com/xbmc/xbmc/issues/19899 - Maybe for here the `NotoSans` about related country needed.
  - This available:
    - NotoSansJP-Regular.ttf
    - NotoSansKR-Regular.ttf
    - NotoSansSC-Regular.ttf
    - NotoSansTC-Regular.ttf
  - Only are them bigger in size
- https://github.com/xbmc/xbmc/issues/15154

Are there also others known?

CastagnaIT commented:
i have this situation, like many other addons i have a multilanguage addon, then can display multiple languages at same time (depends on the type of content) then an example of the situation is this:
- The kodi skin language could be set with any language
- The addon could show at screen e.g. on each listitem label's text mixed language e.g. latin + japanese and latin + korean, etc... at same time

at moment kodi show the missing chars as tofu boxes, and change the skin with arial font, solve the problem enough
this means that is needed the rework mentioned? and this not cover it? (only to understand better)

> The best solution would be to lay down a table in Kodi in which UTF8 character code series is assigned to a font and thus to select each individual character in the appropriate font (similar to Pango or how Chromium does it for websites).

AlwinEsch commented:
Hi @CastagnaIT, for your list item example or the cases where several types are used in a string, a way that adapts it immediately would be ideal. With my way it would currently only be suitable for a special language font and those which in Latin.

Arial is much better in this regard, as it covers a lot more fonts, but unfortunately not all :cry:. Therefore, it would be best to have a way that in all cases a fallback step type is selected which supports the desired type, but also has to be considered that in today's world a mixture can come, e.g. to show URL in it.

So that, like here in the text editor example (everything I found in Kodi for "Programs"), several different ones can be displayed at the same time (and without todo anything by user):
![Bildschirmfoto vom 2021-11-21 16-20-14](https://user-images.githubusercontent.com/6879739/142768256-0cecf7ba-ba03-4aa0-a813-a67c3d2bbfb5.png)

In addition, the user should always get it displayed correctly when he chooses the language and so that he is not forced to do something else, it becomes a deterrent for him, since then only the tofu boxes are shown first.

Just stupid that some types set some standard letters differently e.g. *, .- + _ "0-9
Which then makes it more difficult and its previous and subsequent characters may have to be checked in order to then have the appropriate font in it.

Here as example how I think how can look:
```cpp
struct
{
  uint32_t unicode_code_begin;
  uint32_t unicode_code_end;
  std::string font;
  bool installed;
  std::string name;
  std::unique_ptr<CGUIFontTTF::CFontData> m_font;
} m_fontTable[] = {
  {0x0000, 0x007F, "NotoSans-Regular.ttf", true, "Basic Latin", std::unique_ptr<CGUIFontTTF::CFontData>{}},
  {0x0080, 0x00FF, "NotoSans-Regular.ttf", true, "Latin 1 Supplement", std::unique_ptr<CGUIFontTTF::CFontData>{}},
  {0x0100, 0x017F, "NotoSans-Regular.ttf", true, "Latin Extended-A", std::unique_ptr<CGUIFontTTF::CFontData>{}},
  {0x0180, 0x024F, "NotoSans-Regular.ttf", true, "Latin Extended-B", std::unique_ptr<CGUIFontTTF::CFontData>{}},
  {0x0600, 0x06FF, "NotoSansArabic-Regular.ttf", true, "Arabic", std::unique_ptr<CGUIFontTTF::CFontData>{}},
  {0x0B80, 0x0BFF, "NotoSansTamil-Regular.ttf", true, "Tamil", std::unique_ptr<CGUIFontTTF::CFontData>{}},
};
```
And to create by use. For types not available then downloaded with addon (automatic if needed).
Also should a skin define then in his font xml also something equal (if he want special fonts about)

All UTF8 block numbers are visible here https://www.utf8-chartable.de/unicode-utf8-table.pl

CastagnaIT commented:
so i had understood correctly this is not a solution for addons (the Arial font workaround i already knew)

I think also that the font fallback system should not be something specifically delegated to Skin, because the context is broader,
I haven't analyzed the code so I guess that the loading of font metadata could be missing to determine an automatic search for fonts to use as fallback (also Libass library that now we use to rendering the subtitles has a font fallback system https://github.com/libass/libass/blob/master/libass/ass_fontselect.c#L905-L962)

In my ideal generic view for a future rework:
- The font displayed should fallback (as you said) like the modern OS/Browsers (without depending on skins files)
- The NotoSans font should match in best way possible the languages supported with the Arial font, these additional NotoSans fonts should be included with kodi installation, then without other user interactions.
- Extra languages that are not included within Arial font (and then also in NotoSans) can be added with an additional addon that will provide fonts packages.

Related to this:
- We should move, for ours official Kodi skins (estuary/estuchy), the fonts files in the main "Fonts" folder (where now there is the Arial font) because it makes no sense to maintain a set of duplicate fonts files
- An user should have the possibility to change font skin via GUI (allowing to change with: default skin font, or, Kodi defaults installed fonts, or, custom/extra user installed fonts)
- Emoji EVERYWHERE! just kidding :-) yeah! but actually emoji provide not only designs but sets of very useful icons (like flags, points, color boxes etc) which would be useful in various contexts of addons or to make more pleasant the visualization of some data

The change that this PR makes therefore is a kind of temporary fix for specific Skin case not a real solution,
for me could be ok IMO for skins is not a big change, when the real rework will be done they just need to clean up Font.xml
just that for some people understanding how to configure the XML font fallback with other differents fonts might be complex without a guideline

emveepee commented:
For a user, I recently tried to find NotoSans CJK but couldn't but I dind find Source Hans Sans https://github.com/adobe-fonts/source-han-sans seems much closer to the look of of NotoSans then Arial and it provides many Asian language.  I had to do this renaming Arial perhaps allowing additional ttf fonts could be part of this change.

AlwinEsch commented:
> For a user, I recently tried to find NotoSans CJK but couldn't but I dind find Source Hans Sans https://github.com/adobe-fonts/source-han-sans seems much closer to the look of of NotoSans then Arial and it provides many Asian language. I had to do this renaming Arial perhaps allowing additional ttf fonts could be part of this change.

Will look after it, would be good in any case.

Currently fighting against a stupid mistake in the harfbuzz, hence the temporary ugly commit. Somehow, when English is loaded first, harfbuzz goes crazy when processing e.g. Tamil and its character arrangement does not match.
It's stupid that I can't find anything which from the previous Latin is then used in Tamil and causes the error.
The other way round, if Tamil is loaded first it fits, error somehow comes from the `hb_shape`.

scott967 commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -438,10 +467,54 @@ void GUIFontManager::LoadFonts(const TiXmlNode* fontNode)
+        }
+      }
+
+      //! @todo Why do we tolower() this shit?
```
Maybe improve this comment a bit?

AlwinEsch replied:
Yeah, was already in https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIFontManager.cpp#L441

Thinking of looking at the problem and removing the comment: smile:.

scott967 commented:
Issue #19899 is correct, in that the unicode consortium created a set of "unified" ideographs intended to merge various language usages into a reduced set of codepoints.  But as indicated in the issue, a language-specific font can use a different glyph for a codepoint.  (Also for older Vietnamese typography but probably not required to support).

Noto Sans is the only font family I know that is licensed under SIL and provides fonts with glyphs for each of the languages.

I think consideration should be given to using ISO 15924 script tags vice our current practice using language which can be ambiguous due to our main use of ISO 639-2/B  language.  So for the East Asian case Hans and Hant for various Chinese dialects but also Jpan and Kore would allow for selection of proper glyphs (fonts) for the unified CJK codepoints (generic Hani).

For the fallback font concept, I don't know how freetype2 deals with fonts using different units per em (1000 or 2048).

In addition, since implementation of Harfbuzz shaping engine, the shaping tables for Latn are different between the basic Noto Sans-Regular and the various Noto Sans CJK fonts.  Not sure how that is taken into account when rendering?

AlwinEsch commented:
Thanks @scott967 for the infos.

Have also becomed it now working with a:
```cpp
std::vector<CFontTable> CGUIFontTTF::m_fontFallbacks = {
  {0x0000, 0x024F, "NotoSans-Regular.ttf", true, " Latin"},
  {0x0600, 0x06FF, "NotoSansArabic-Regular.ttf", true, "Arabic"},
  {0x0B80, 0x0BFF, "NotoSansTamil-Regular.ttf", true, "Tamil"},
  {0x0C00, 0x0C7F, "NotoSansTelugu-Regular.ttf", true, "Telugu"},
};
```
see temporary commit.

Need to clean it up now and to make performace test, hope it makes not so big differences.

AlwinEsch commented:
Only does the text length not fit Kodi's screen size, but thats not from me :smirk::rofl: 
![Bildschirmfoto vom 2021-11-23 04-58-11](https://user-images.githubusercontent.com/6879739/142969507-33c5603e-4b0e-4993-85e2-6ece4b0778de.png)

AlwinEsch commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -54,6 +54,173 @@
+    {0xFE50, 0xFE6F, "NotoSansSC-Regular.ttf", true, CFontTable::ALIGN::LTOR, "Small Form Variants"},
+    {0xFE70, 0xFEFF, "NotoSansArabic-Regular.ttf", true, CFontTable::ALIGN::LTOR, "Arabic Presentation Forms-B"},
+    {0xFF00, 0xFFEF, "NotoSansSC-Regular.ttf", true, CFontTable::ALIGN::LTOR, "Halfwidth and Fullwidth Forms"},
+    {0xFFF0, 0xFEFF, "NotoSans-Regular.ttf", true, CFontTable::ALIGN::LTOR, "Specials"},
```
Have now added the mostly complete table with set of unicodes and reference to related font.

Would look also to implement unicodes with more as 16 bit and to allow emojis with it.

The selection as worked good during test, only need to look to become better alignment, some fonts are a bit lower or upper as the normal Latin words. Also was in Arabic the alignment not to right.

scott967 commented:
I compared Noto Sans-Regular.ttf  that is being removed from Estuary skin fonts to the new Noto Sans-Regular.ttf that is being placed in media/Fonts

There are many difference between these two font files.
From a technical standard, the current font file has a UPM of 2048 while the new one is 1000. 
From a glyph / language coverage the current font has 3690 glyphs while the new one only 3271.  In attempting to analyze the difference I see that the new font removes support for Armenian, Hebrew, and Coptic scripts.  Looking further I see that the glyphs for Unicode blocks for Armenian, Hebrews, Coptic are removed, as are a great many symbols / math operators, and the like.

Was this intended?

lrusak commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -54,6 +54,173 @@
 
 #define TAB_SPACE_LENGTH 4
 
+std::vector<CFontTable> CGUIFontTTF::m_fontFallbacks = {
```
anonymous namespace instead of a static member?

lrusak commented on xbmc/guilib/GUIFontTTF.h:
```diff
@@ -148,12 +192,16 @@ class CGUIFontTTF
+  float m_aspect{1.0f};
+  float m_lineSpacing{1.0f};
+  bool m_border{false};
   std::string m_strFilename;
```
Not changed but is this actually used for anything? I see we had a duplicate below that was slightly different `m_strFileName` which was used by the getter `GetFileName()` but this doesn't seem to be used by anything...

AlwinEsch replied:
On other request becomes them removed.

Yes makes no sense to use as value, if nowhere used.

lrusak commented on xbmc/guilib/GUIFontTTF.h:
```diff
@@ -136,7 +180,7 @@ class CGUIFontTTF
   float GetCharWidthInternal(character_t ch);
   float GetTextHeight(float lineSpacing, int numLines) const;
-  float GetTextBaseLine() const { return (float)m_cellBaseLine; }
+  float GetTextBaseLine() const { return (float)m_fontMain.m_cellBaseLine; }
```
c++ style cast

lrusak commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -281,50 +451,64 @@ bool CGUIFontTTF::Load(
-  m_cellHeight   = cellAscender - cellDescender;
+  data.m_cellBaseLine = cellAscender;
+  data.m_cellHeight = cellAscender - cellDescender;
+  fprintf(stderr, "data.m_cellHeight %i %i\n",data.m_cellHeight, data.m_cellBaseLine);
```
remove?

AlwinEsch replied:
Still temporary, also not finished the cleanup. Thought on begin to allow only fallback to one font and not thought to become working with all :sweat_smile:.

Planned to make a separate request and to cleanup the whole Fontsystem before here comes in (without work way changes). There are much parts inside the font source files where need really a cleanup, think them from begin (maybe ~2006) and not really changed since them. Specially his singleton way with `XBMC_GLOBAL_REF(CFreeTypeLibrary, g_freeTypeLibrary);` :smirk:.

lrusak commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -640,21 +828,21 @@ float CGUIFontTTF::GetCharWidthInternal(character_t ch)
 float CGUIFontTTF::GetTextHeight(float lineSpacing, int numLines) const
 {
-  return (float)(numLines - 1) * GetLineHeight(lineSpacing) + m_cellHeight;
+  return (float)(numLines - 1) * GetLineHeight(lineSpacing) + m_fontMain.m_cellHeight;
```
c++ cast

lrusak commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -931,13 +1199,13 @@ bool CGUIFontTTF::CacheCharacter(wchar_t letter, uint32_t style, Character* ch,
   ch->letter = letter;
   ch->offsetX = (short)bitGlyph->left;
-  ch->offsetY = (short)m_cellBaseLine - bitGlyph->top;
+  ch->offsetY = (short)m_fontMain.m_cellBaseLine - bitGlyph->top;
```
c++ cast

lrusak reviewed (commented):
To be honest this was pretty hard to read. The font code in general is quite a mess (not your fault it's just legacy code). I've commented in a couple places where I think some changes can be made but overall it's too hard to understand the changes by looking at the diff.

I think it would be beneficial (after this PR) for someone to do a cleanup of the font code. I started a bit but didn't get too deep as I was working on some other stuff.

Possible future considerations include:
 - CreateGUIFontTTF returning a std::unique_ptr (or possibly std::shared_ptr if required)
 - `m_vecFontFiles` and `m_vecFonts` could hold a smart pointer from above
 - Removing the friend class of `CGUIFont` and use better inheritance.

Overall It's a big improvement to our font handling and should help many languages. Great job!

AlwinEsch commented:
> I compared Noto Sans-Regular.ttf that is being removed from Estuary skin fonts to the new Noto Sans-Regular.ttf that is being placed in media/Fonts
> 
> There are many difference between these two font files. From a technical standard, the current font file has a UPM of 2048 while the new one is 1000. From a glyph / language coverage the current font has 3690 glyphs while the new one only 3271. In attempting to analyze the difference I see that the new font removes support for Armenian, Hebrew, and Coptic scripts. Looking further I see that the glyphs for Unicode blocks for Armenian, Hebrews, Coptic are removed, as are a great many symbols / math operators, and the like.
> 
> Was this intended?

Mmmh, think to has changed by mistake where taken from https://github.com/googlefonts/noto-fonts.

Then I'm not sure how it could the best, leave in our from before, or to take from the git as more updated?

AlwinEsch commented:
Should now be largely through and done. Will then update the other request and then clean up here thoroughly in order to have correct commits and clang format.

I will try to support the text for "RTOL" where it is then also placed on the right in another request and the support of Emoji's in a new one (will be more work to process the graphic textures).

Due to the loading of the associated add-on, the following now appears:
![Bildschirmfoto vom 2021-11-28 18-16-24](https://user-images.githubusercontent.com/6879739/143779361-a4ae0853-9b0c-48c2-bbc7-a43359005b27.png)

CastagnaIT commented:
I don't know about the others but imo i would change the text on the dialog:
ID 2105: `Missing written language support`
maybe better: `Missing font language support` ?

and instead for
ID 2106: `Install addon "resource.language.hi_in" with font to support written language "Devanagari"?`
i don't think the final user give importance to addon code "resource.language.hi_in",
maybe more user friendly something like:
`Do you want to install the font to support "Devanagari (hi-in)" language?`
or
`Install font to support "Devanagari (hi-in)" language?`

AlwinEsch commented:
Is now updated and cleaned up after the other cleanup request came in.

I have also updated the request here https://github.com/xbmc/repo-resources/pull/190 and also contains the optional font addons, in case someone e.g. wants to have Egyptian hieroglyphs displayed.

CastagnaIT commented on xbmc/utils/ColorUtils.h:
```diff
@@ -19,6 +19,7 @@ namespace COLOR
 {
 
 typedef uint32_t Color;
+typedef uint32_t ColorSelect;
```
what is the difference between `Color` and `ColorSelect`?

AlwinEsch replied:
There is currently a mixture about two ways to get the color.
One with `ARGB` format and the other with normal integer number to get color within list, think can be better to declare as `unsigned int` instead of `uint32_t`.

CastagnaIT replied:
but `ColorSelect` type is passed to `character_t` struct this last one use `Color` type argument and then the value is stored/converted again as `ColorSelect` type. In other places the `character_t` value is passed with `Color` type value, all this seems a bit inconsistent and mixed.
Then `ColorSelect` is only a index number for a list, that has nothing to do with a color, why not call it something like `ColorIndex` ?

AlwinEsch replied:
I updated it and now uses `typedef unsigned int ColorIndex;` (really better with index).
I also used `constexpr ColorIndex INDEX_DEFAULT = 0;`, fits better than `UTILS::COLOR::NONE`.

CastagnaIT replied:
now is much better i like it
just remember to change `UTILS::COLOR::Color` to `UTILS::COLOR::ColorIndex` in to `character_t` constructors

AlwinEsch commented:
To make checks I uploaded also now my python script about to here: https://github.com/AlwinEsch/script.fonttest
The `example.txt` can be edited to what to check.
Another way to check it too, is by switching Kodi's languages.

Only aware that on font download, it not works correct now, as https://github.com/xbmc/repo-resources/pull/190 not in, need to set by hand from there. 
E.g. clone https://github.com/AlwinEsch/repo-resources/tree/Nexus-change and to copy to `~/.kodi/addons`.

scott967 commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -51,12 +61,439 @@
+     {CFontTable::ADDON::RESOURCE_FONT_LIMITED, "resource.font.limited"},
+     {CFontTable::ADDON::RESOURCE_FONT_EXCLUDED, "resource.font.excluded"}}};
+
+// List about availabe characters:
```
spell  check

List about available characters:

scott967 commented:
Tested this and I think it is going to need some good user support info.  I added a music source, and got the popup "Do you want to install the font to support "Halfwidth and Fullwidth Forms" language?"  I understand why I got that and what it means, but I'm not sure most users would.  I'm thinking "characters" might be better than "language", but maybe I am bike shedding.  Though maybe it would be better to ask the user if they want to install the actual language resource addon that contains the needed font, in this case resource.language.zh_cn (Chinese simplified/Hans).  (I couldn't test this too much, because it downloaded the resource from kodi repo that doesn't include a font file.  I don't know any way to force it to get the updated resource locally).

AlwinEsch commented:
Maybe as this on dialog?:
```
Do you want to install the add-on to support characters?

Add-on: resource.language.zh_cn (Chinese simplified/Hans)
Character Typ: "Halfwidth and Fullwidth Forms" 
```

scott967 commented:
Something like that I think users should understand.  It probably only affects few users, but for those I think it helps to prevent confusion.

AlwinEsch commented:
Have updated and now brings in dialog as this:
![Bildschirmfoto vom 2021-12-30 00-42-35](https://user-images.githubusercontent.com/6879739/147711044-ead8c9d0-c397-45ef-8019-ca2e83ae7874.png)

enen92 commented on xbmc/addons/AddonInstaller.cpp:
```diff
@@ -189,22 +189,23 @@ bool CAddonInstaller::InstallModal(const std::string& addonID,
 {
   return DoInstall(dependsId, repo, BackgroundJob::NO, ModalJob::NO, AutoUpdateJob::NO,
-                   DependencyJob::YES, AllowCheckForUpdates::YES);
+                   DependencyJob::YES, AllowCheckForUpdates::YES, -1);
```
Can this be changed to an enum (like the other arguments in this method) at least for consistency and avoid the magic number?

AlwinEsch replied:
Wouldn't be possible as directly as it is really a little magic number.
The associated add-on type must define its appropriate values.
I will use it as default with the:
```cpp
constexpr int ADDON_INSTALL_OPTIONAL_SELECT_UNDEFINED = -1;
```
Ah idea about a better name (where not so long, but still understandable)?

enen92 commented on xbmc/addons/LanguageResource.cpp:
```diff
@@ -92,14 +92,19 @@ bool CLanguageResource::IsInUse() const
+  if (optionalTypeDependend == POST_INSTALL_LANGUAGE_RESOURCE_NO_SELECT)
+  {
+    // Force to reload current strings
+    g_langInfo.SetLanguage("", true);
```
maybe this can be defined as a new method in `g_langInfo`? Without the comment it's hard to understand the reason for this call (or what its intended behavior is)

AlwinEsch replied:
Has taken from:
https://github.com/xbmc/xbmc/blob/master/xbmc/Application.cpp#L4677-L4679

Maybe I can add a function above the other as?:
```cpp
void CLangInfo::ReloadCurrentLanguage()
{
  SetLanguage("", true);
}

bool CLangInfo::SetLanguage(std::string language /* = "" */, bool reloadServices /* = true */)
{
...
```

Only not sure if it makes really sense if only called by one place.

scott967 commented:
I built this with the new commits.  I think the dialog strings are OK.  But I had a problem in test.  Note that I can't figure out a way to force Kodi to get the updated language resource addons -- it always pulls from the official mirror.

After a fresh install I changed language to Japanese.  The language changed (so I got the ,notdef tofu) and I got the dialog asking me to install Chinese (simpilified).  That was expected, as Kodi installed the current release resource.language.jp_JP which doesn't have the font.  If I clicked "no" I just got the dialog popped up again.  I guess that isn't a surprise.  I couldn't exit Kodi via UI so I killed it in the debugger.  Then I manually replaced the language resource with the updated one that has NotoSansJP-Regular.ttf included.  I thought that when I restarted Kodi that would solve it.  But it didn't use the new font.  I still get the tofu and the dialog to install simplified Chinese.

I don't know if this is a side effect of manually updating the resource so something else?

What I was expecting was if user switches UI language to Japanese, Korean, or traditional Chinese, that the specific font for that variant is used rather than simplified Chinese.

AlwinEsch commented:
> What I was expecting was if user switches UI language to Japanese, Korean, or traditional Chinese, that the specific font for that variant is used rather than simplified Chinese.

With CJK is it really hard to manage the best, on the Chinese `NotoSansSC-Regular.ttf` are the most characters inside defined, also most for Japan and South Korea. To some special unicode's it takes then the others. Also found in test that on Japanese text few characters was missing, where then was available in the Chinese.

It seems to me that when setting the character IDs, they could not coordinate correctly and different ones in one ID series. Especially the one from 0x020000 to 0x03134F which made me crazy.

Wouldn't avoid installing Chinese even if Japanese was set as the language.

But maybe it would be good to bring in the request for language add-ons soon so that at least their fonts are already available and correct in updates.

AlwinEsch commented:
Hi @scott967, have added a fix the prevent the deadloop if only "no" becomes selected and added an additional dialog to show.
With the Text I'm not really like on new dialog as it is really long:

```
Selected language add-on for specifying the font does not 
always correspond to the language used itself.
[f you  choose "no", this install is interrupted in the currently
running Kodi, but after a restart and the font is needed again, it 
is brought up again!
Are you sure to cancel?
```

Another idea about?

Also comes it after Kodi restart again (if relates to Skin languge itself, e.g. the Japanese), to prevent forever is not good for me as it can then on other cases bring problems.

![Bildschirmfoto vom 2022-01-04 17-13-41](https://user-images.githubusercontent.com/6879739/148089960-660e1b9e-acea-40a0-90fc-85674e367100.png)

AlwinEsch commented:
Perhaps there would also be a possibility with the CJK addons resource.language.zh_tw and resource.language.ja_jp to add resource.language.zh_cn as a dependency?

scott967 commented:
Built this today (1/17).  When I switch UI language to Japanese, I still get the tofu boxes on the yes/no buttons in the dialog (asking to load simplified Chinese).  I still think it is better to first try to get characters from the Japanese font and just ask to load simplified Chinese if the chars are not available in the Japanese font.  (Same thing if switching UI language to Chinese (traditional).  It isn't likely an issue of Korean, as Hangul chars load the Korean font anyway.

I then did some testing with UI in English.  I get  a repeatable crash in Windows VS 2017 debugger.
What I did is go Music - Files - add source - browse and navigate through folders to one with folder names in Korean and Japanese.  First I get dialog to install zh_cn due to halfwidth and fullwidth forms.  I clicked "YES" and Chinese (simple) language was installed (from a repo I created containing the new resource.language addons -- not the default Kodi repo) then I got a second dialog to install ko_kr due to hangul syllables.  I clicked "YES" and Kodi threw an exception in CGUIControlLookup::RemoveLookup() for an access violation vcruntime140d.dll.   I think the problem is in going into DialogMediaSource and then FileBrowser where Kodi needs the additional fonts, the yesno dialog gets loaded.  After selecting YES the zh_cn language is loaded, all the windows are unloaded and skin is reloaded.  Then the music window gets reloaded and yesno dialog gets loaded again, but this time the file browser and mediasource dialogs haven't been opened.  My guess is this attempt to load a second language after the file browser has been closed is causing a problem that leads to the exception.

CastagnaIT commented:
@AlwinEsch the NotoSans-Regular.ttf that come from skin estuary folder is not the original google font
and has been patched to add some additionals chars that seem to me that you have already included here with the font addons and/or additionals ttf's (like Noto Sans Symbols ttf)

suggest to restore the original google version of NotoSans-Regular.ttf and drop the customized one,
i also noticed that in the font preview the customized one looks weird,
on the left the original google, on the right the kodi customized:
![immagine](https://user-images.githubusercontent.com/3257156/150347388-dfc7e7ef-f150-4aa8-983d-99c1a296be0b.png)

AlwinEsch commented:
@scott967 
I'll clean it up in the next few days, maybe it would really make sense to put the CJK fonts in a separate addon, so that the respective string addons have a dependency on a single associated font addon.

@CastagnaIT 
Will then reset Kodi's "NotoSans-Regular.ttf" to the original google version. Thought so too at first, but wasn't sure if it was desirable or good.
Funny that it is displayed correctly in Kodi and so different in the font editor like yours.

CastagnaIT commented:
the modified version of "NotoSans-Regular.ttf" in kodi, has languages that you have already add in the fallbacks
should not appear regressions to use the original version

yol commented:
Cool stuff! Would really like to see this work! I think these problems has significantly heightened the barrier to entry for Asian users for years.

AlwinEsch commented:
Have done an rebase. Needs cleanup on some places and few fixes. Inside "Arabic" the ASCII charactes show wrong now.
Mainly it still works.
Hopefully I becomes it ready soon :crossed_fingers:.

AlwinEsch committed:
[LanguageCodes] add func. CLangInfo::ReloadCurrentLanguage()

In relationship to coming support about use of several different fonts and to make e.g. a reload after an font add-on was installed.

AlwinEsch commented:
Played a bit with it and also now allow align right to left on related written languages.
![Screenshot_20220905_191316](https://user-images.githubusercontent.com/6879739/188494066-8713e0b6-c840-4e12-9a21-0f0f7ce70c03.png)

AlwinEsch committed:
[gui][fonts] move NotoSans-Regular.ttf and add new ones too

With this, NotoSans-Regular.ttf is moved to "media/Fonts" in order to have it available on all skins. The background is that Noto Fonts are available on all fonts in the world.

For this reason, other new Noto fonts are also being introduced to allow common symbols and expressions.

All fonts that belong specifically to a language are stored in "resource.language.*", Or in the case of special ones in "resource.font.*" Addons.

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -651,163 +1258,177 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
 
   // letters are stored based on style and glyph
-  character_t ch = (style << 16) | glyphIndex;
+  const glyph_and_style_t ch =
```
I do not understand much this change: `style` is in range 0 to 6 and `glyphIndex` 0 to 65536 because TT / OpenType fonts can only store 65536 glyphs (AFAIK). Then to map this in one variable should be:

```
style           ch (depending of glyphIndex)
 0 --------->       0 to 65,536
 1 --------->  65,637 to 131,072
 2 ---------> 131,073 to 196,609
 3
 4
 5
 6 ---------> 393,216 to 458,752
```
And the formula to accomplish this is:

ch = (style * 65536) + glyphIndex

If we want use bits shift method (because is supposed is more optimized code):

`const uint32_t ch = (style << 16) | glyphIndex;`       

(2^16 = 65536)

And this is what was there before : D

I also think it's debatable whether it's necessary to use bit shifting instead of multiplication since it's supposed to be the compiler's job to optimize the code for each platform and there's no use wanting to do it manually because the generated assembly code will probably be the same.

It is even possible that the compiler will do better if we express what we want to do and not how to do it.

Also is not need use `int64_t` for `glyph_and_style_t` because max range is 458,752 or maybe I'm not taking something new into account?

fuzzard replied:
Just as an FYI, there is a spec proposal to remove the 65k glyph limit. https://github.com/harfbuzz/boring-expansion-spec

AlwinEsch replied:
Yes, that's seen during search on web. A 16bit 0xFFFF can no more enough for all cases.

AlwinEsch replied:
About the shift, yes is better to use an "<< 32", thought first to define the size as this to have more clean on changes, but makes no sense ☺️.

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -651,163 +1258,177 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
+                                     CFontData* fontData)
 {
-  const wchar_t letter = static_cast<wchar_t>(chr & 0xffff);
+  typedef uint32_t char_quik_t;
```
typedef seems unnecessary as only is used in two places

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -651,163 +1258,177 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
+  if (glyphIndex < MAX_GLYPH_IDX && fontData->m_mainFont)
   {
-    character_t ch = (style << 12) | glyphIndex; // 2^12 = 4096
+    char_quik_t ch = (style << 12) | glyphIndex; // 2^12 = 4096
```
Maybe `const uint32_t` here and in the other place that is used

AlwinEsch replied:
Can change it, thought first to have more clean, as the "ch" not direct understandable and stores the character a bit different with this flags.

AlwinEsch replied:
Have changed to `const uint32_t`

AlwinEsch committed:
[gui][fonts] use right to left alignment on related fonts

This adds the initial part to handle written text from right to left.
This is the initial try about. About spaces, tabs and defined gui text widths it becomes hard to handle correct in all conditions.
For now it makes by this languages the align to opposite and render the characters backward, where needed to have in sizes correct.

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -923,16 +1674,16 @@ bool CGUIFontTTF::CacheCharacter(FT_UInt glyphIndex, uint32_t style, Character*
 
   // set the character in our table
-  ch->m_glyphAndStyle = (style << 16) | glyphIndex;
+  ch->m_glyphAndStyle = (static_cast<decltype(ch->m_glyphAndStyle)>(style) << 32) | glyphIndex;
```
Seems more clear:

`ch->m_glyphAndStyle = (static_cast<glyph_and_style_t>(style) << 32) | glyphIndex;`

AlwinEsch replied:
Have changed, was gone down where I changed the other. By other place was come a compile warning with it.

AlwinEsch committed:
[gui][fonts] Pass characters in structure

With this, the characters are transferred within a structure in which the color and style are then stored separately.

This has the background to get 32bit UTF codes usable as characters and to omit the current shifts which previously put the type in character itself.

enen92 commented on xbmc/addons/Addon.cpp:
```diff
@@ -501,9 +501,9 @@ void OnPreInstall(const AddonPtr& addon)
 }
 
-void OnPostInstall(const AddonPtr& addon, bool update, bool modal)
+void OnPostInstall(const AddonPtr& addon, bool update, bool modal, int optionalTypeDependend)
```
maybe some enum class can be used here instead?

AlwinEsch replied:
As enum not the best for there, as it can come thats an other add-on type can use this in future too, but there e.g. as integer. That's make it hard to have usable in all possible conditions.

garbear replied:
I'd consider reframing the problem with an enum, because it took me a while to track down `POST_INSTALL_LANGUAGE_RESOURCE_NO_SELECT` as the only non-negative option. What does zero mean?

AlwinEsch replied:
Changed to use an `enum` for it.

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -4831,7 +4831,31 @@ msgctxt "#2104"
 msgstr ""
 
-#empty strings from id 2105 to 2202
+#. Heading text for yes/no dialog to ask for add-on install to get his font file
```
Can you reword this? It requires a deep understanding of how font language support works, might confuse translators.

AlwinEsch replied:
Have done a reword for here and on others.

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -4831,7 +4831,31 @@ msgctxt "#2104"
+msgid "Missing font language support"
+msgstr ""
+
+#. Used to show notification about automatic language add-on installation, for here the used text
```
Can you phrase this more in a way that describes the question being asked in the string? Also document what each {} means.

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -4831,7 +4831,31 @@ msgctxt "#2104"
+# Notification message if add-on installation about font failed.
+#: xbmc/guilib/GUIFontTTF.cpp
+msgctxt "#2107"
+msgid "Installation of the add-on \"{0:s}\" to support the written language \"{1:s}\" failed."
```
Do we distinguish between written languages and spoken ones elsewhere in the GUI? This is the first time I'm seeing "written" describing a language.

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -4831,7 +4831,31 @@ msgctxt "#2104"
+#. Used to show notification about automatic language add-on installation, to report result by cancel.
+#: xbmc/guilib/GUIFontTTF.cpp
+msgctxt "#2108"
+msgid "Selected language add-on for specifying the font does not always correspond to the language used itself.[CR]If you choose "no", this install is interrupted in the currently running Kodi, but after a restart and the font is needed again, it is brought up again![CR]Are you sure to cancel?"
```
This string crosses my github width twice over, seems like it would create a wall of text. Hint: people don't read. Try to give a few words that lets someone get through the dialog quickly.

AlwinEsch replied:
Have removed this dialog show and the text.

garbear commented:
I left some comments on the GUI strings. Not an issue with the current state, an issue with my understanding of multifont.

AlwinEsch committed:
[addons] allow optional value on OnPostInstall calls

The background to this is to give each add-on type the opportunity to have its own value for an installation and to be able to identify any additional work.

E.g. in the upcoming changes to avoid changing the Kodi language in the font installation using the language addon.

AlwinEsch commented:
Have added a new setting to make easier for user and to prevent the dialog shows for fonts (if not set in setting).

This now added with last commit and new:
![Screenshot_20220910_142617](https://user-images.githubusercontent.com/6879739/189483621-f2b13133-4f46-4964-966b-fd022310dbad.png)
![Screenshot_20220910_142635](https://user-images.githubusercontent.com/6879739/189483626-b8023162-d037-4051-8a6d-047746841a0f.png)

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -822,47 +1452,180 @@ CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr, FT_UInt glyph
 
-      if (ch < LOOKUPTABLE_SIZE)
-        m_charquick[ch] = m_char.data() + i;
+      if (ch < LOOKUPTABLE_SIZE && fontData->m_mainFont)
```
If I'm not wrong `m_charquick` is only used with main font. If so, it would be better not execute this for loop at all (if font is not main)

```c++
if (fontData->m_mainFont)
{
  // update the lookup table with only the m_char addresses that have changed
  for (...
}
```

AlwinEsch replied:
You are right, yes to make the `if` before is much better and changed it.

garbear commented on xbmc/addons/IAddon.h:
```diff
@@ -27,7 +27,11 @@ using AddonInstanceId = uint32_t;
 {
   //! Default value if unused on add-on type.
-  UNUSED = -1
+  UNUSED = -1,
```
Is -1 required here? Can it be omitted and left as 0 to avoid mixing enums and ints, or is -1 important?

garbear commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -60,9 +70,449 @@ constexpr int GLYPH_STRENGTH_BOLD = 24;
 constexpr int GLYPH_STRENGTH_LIGHT = -48;
 constexpr int TAB_SPACE_LENGTH = 4;
 
+constexpr std::chrono::hours JOB_RETRY_TIMEOUT{24};
```
I'm not that familiar with C++ units, but can you initialize with `24h` to be type-correct?

AlwinEsch replied:
Checked, not works with it and will leave as 24.

garbear commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -60,9 +70,449 @@ constexpr int GLYPH_STRENGTH_BOLD = 24;
 constexpr int TAB_SPACE_LENGTH = 4;
 
+constexpr std::chrono::hours JOB_RETRY_TIMEOUT{24};
+constexpr std::chrono::hours JOB_CANCEL_TIMEOUT{0xFFFFFFFF};
```
Can this be set to `std::numeric_limits<std::chrono::hours>::max()`?

AlwinEsch replied:
Yes thats work, cool and thanks!

garbear commented on system/settings/settings.xml:
```diff
@@ -3791,6 +3791,18 @@
         </setting>
+        <setting id="locale.fontinstall" type="integer" label="2108" help="2109">
+          <level>1</level>
+          <default>1</default> <!-- InstallWay::NO_ASK -->
```
You can use `InstallMethod` instead. "Way" means pretty much anything but "method" means that a particular procedure is used.

AlwinEsch replied:
Have changed to "method", also used it as this on the used strings.po.

garbear commented:
Getting closer! Only found a few minor issues left.

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -651,163 +1270,174 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
+    {
+      const CFontTable* fontTable;
+      fontData = FindFallback(chr.letter, fontTable);
+      if (!fontData || (glyphIndex = FT_Get_Char_Index(fontData->m_face, chr.letter)) != 0)
```
I think this should be `== 0` 

`FT_Get_Char_Index(fontData->m_face, chr.letter)) == 0`

`FT_Get_Char_Index` error is when index is zero

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -651,163 +1270,174 @@ std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecTe
+      if (!fontData || (glyphIndex = FT_Get_Char_Index(fontData->m_face, chr.letter)) != 0)
+      {
+        CLog::Log(LOGERROR,
+                  "CGUIFontTTF::{}: Unable to find font about character {:x}, font data found: {}",
```
Maybe use `CLog::LogF` style call as others log calls in this file are in this format

AlwinEsch committed:
[gui][fonts] Allow other fonts if not available in the main font

This introduces the support to use several different fonts at the same time.
This makes it possible to have Kodi compatible with all available types, regardless of what skin addon defines.
If the required font is not available, the user is asked to install a corresponding add-on.

AlwinEsch committed:
[gui][fonts][settings] allow font install method selection

By this becomes an new setting added where allow the use to select the method how install done.
Values are:
- Never: User becomes never asked and install never done
- Always, if not available: If not available, becomes it automatic installed without user confirmation
- With user confirmation: User becomes asked about

AlwinEsch reviewed (commented):
Thanks @garbear and @thexai, have changed on your notes.

CastagnaIT commented on xbmc/guilib/GUIVecText.h:
```diff
@@ -0,0 +1,51 @@
+
+struct character_t
+{
+  character_t(char l, int s, UTILS::COLOR::Color c)
+    : letter(static_cast<char32_t>(l)), style(s), color(c)
+  {
+  }
+  character_t(char16_t l, int s, UTILS::COLOR::Color c)
+    : letter(static_cast<char32_t>(l)), style(s), color(c)
+  {
+  }
+  character_t(wchar_t l, int s, UTILS::COLOR::Color c)
+    : letter(static_cast<char32_t>(l)), style(s), color(c)
+  {
+  }
+  character_t(char32_t l, int s, UTILS::COLOR::Color c) : letter(l), style(s), color(c) {}
+
```
here in character_t struct there are mixed types
`UTILS::COLOR::ColorIndex` VS `UTILS::COLOR::Color`
is intentional?

CastagnaIT commented:
reopened, this should be one day implemented is quite important

garbear commented:
> reopened, this should be one day implemented is quite important

All my PRs and issues I was tracking were closed after https://github.com/xbmc/xbmc/pull/26763. Cheap to reopen.

CastagnaIT commented:
> > reopened, this should be one day implemented is quite important
> 
> All my PRs and issues I was tracking were closed after #26763. Cheap to reopen.

should be exist a flag to prevent a future "Stale and close" otherwise PR will be lost

garbear commented:
> should be exist a flag to prevent a future "Stale and close" otherwise PR will be lost

Yes, I manually went through my email and reopened all PRs and issues I didn't want lost, but maybe there's a better way.

scott967 commented:
A big concern I have without digging in is how does HarfBuzz shaping work under this condition?  Kodi Arial does have arabic script but not sure about the shaping aspect of either Arial or your fallback design.  Kodi Arial is missing some scripts (Indic languages plus any outside the BMP like emojis).  What happens in those cases?  Back to the current method?  I also have experience in font mash-up that differences in m-box/ascenders/descenders/metric can cause render issues (text no longer fitting in skin defined space).  Have you looked at that?

axbmcuser commented:
@ihipop

Great idea as an interim solution! :-)
Do you have a link to the source of the patch so it could do some testing with it?
Thanks

ihipop commented:
> @ihipop
> 
> Great idea as an interim solution! :-)
> Do you have a link to the source of the patch so it could do some testing with it?
> Thanks

@axbmcuser

I've tested this on my local LibreELEC build for a couple of months and it's been running smoothly without any unexpected crashes. I'd be happy to prepare a PR if this approach is welcome (just checking first so I don't waste any effort). If you have other thoughts or want to see what I've done, you can check it out here: 

https://github.com/ihipop/EnhancedELEC/blob/5ec58e7354cfe455fa38cba968d872fe359ab067/packages/mediacenter/kodi/patches/kodi-100.90-cjk-font-fallback.patch

axbmcuser commented:
@ihipop 

Thanks for the source!
Will also try this on my personal CoreELEC build and see how it goes and what maybe comes to mind while using it.

I created an extensive custom font support solution (chinese simple, chinese traditional, japanese, korean, hebrew, arabic, thai and hindi) for [Confluence ZEITGEIST](https://forum.kodi.tv/showthread.php?tid=242892) and while it technically works very well on the skin side (including fully automatic font switching by a helper python add-on which works together with skin), i never released it because i wasn't satisfied with the wonky situation/results/downsides in regards to merged fonts. (despite merging some myself with somehow okay-ish results)

I will compare how this fallback solutions feels visually in comparison and report back. Cheers

axbmcuser commented:
@ihipop 

Haven't had the time yet to compile a new build.

But here already some food for thoughts & questions:

- Did you ever try to use a different fallback font than Kodis stock "arial"?
- What about slightly extending your nice idea like this:

`<font>`
`	<name>Font15</name>`
`	<filename>MyFont.ttf</filename>`
`	<fallback>FallbackFont.ttf</fallback>`
`</font>`

This would instantly further multiply the usefulness of your solution from my pov.
What do you think & would you be up to looking into it?

Thanks again! :-)

ihipop commented:
@axbmcuser 

The reason we are seeing so many "tofu" in current Kodi skins is that the built-in fonts provided by many skin authors lack full character set support. My current thought is to use Kodi's stock "arial" font to ensure basic usability, so that at least the tofu blocks won't appear on the UI.

If the configuration you mentioned requires setting fallback fonts on a per-skin basis, then the initiative to resolve this issue would still lie with the skin authors. In that case, if a skin author doesn't update their skin to support this, wouldn't users be completely unable to benefit from this feature?

I am more than happy to discuss your idea further, but I am also eager to submit a minimal fallback solution first to resolve the fundamental usability problem. Would introducing a new skin extension configuration delay the progress of [merging the  PR in planning](https://github.com/ihipop/EnhancedELEC/blob/5ec58e7354cfe455fa38cba968d872fe359ab067/packages/mediacenter/kodi/patches/kodi-100.90-cjk-font-fallback.patch) ? Would this approach be accepted by the Kodi team? I hope someone from the official Kodi team can evaluate and provide some guidance on this proposal. If necessary and permitted, I can open a new issue specifically to discuss this expanded feature.

axbmcuser commented:
@ihipop 

Wasn't referring to adding mandatory requirements your solution and understand the idea wanting to keep it as simple as possible.

Still i think 2 points could be relevant:

1) Merging your - even if great - simple solution (which is instantly forced on by default with no choice to turn it off, not even for debugging) could maybe be accompanied by minor cause for hesitation

2) Refining the solution in a still-as-simple-as-possible way would avoid 1) and multiply the usefulness and also would make it possible to soft-instroduce the solution, further countering potential hesitation to merge it:


(Updated refinements)

Placing things in hands of fontset level like this:


**Example 1**, fallback enabled by choice:

`<font>`
`	<name>Font15</name>`
`	<filename>MyEnglishMainStreamFont.ttf</filename>`
`	<fallback>true</fallback>`
`</font>`

Result:
Fallback to Kodis stock arial.ttf active for missing characters


**Example 2**, fallback enabled by choice with defining a fallback font file:

`<font>`
`	<name>Font15</name>`
`	<filename>MyEnglishMainStreamFont.ttf</filename>`
`	<fallback>true</fallback>`
`	<fallbackfilename>FallbackFont.ttf</fallbackfilename>`
`</font>`

Result:
Fallback to optionally defined "FallbackFont.ttf" active for missing characters

ihipop commented:
@axbmcuser
I see.

That's a great suggestion. I'll get right on it and implement the fallback font as an additional feature in the Kodi font settings. I was wondering, can I open a dedicated PR for this? Is that allowed?

AlwinEsch committed:
reverts2

AlwinEsch committed:
fix parts, becomes on coming cleanup included to other commits

AlwinEsch commented:
Is rebased now, but not cleaned up and not runtime tested! 

**There's a very high probability that it won't work yet!**

Some parts where was come in to Kodi are temporary now reverted and becomes on coming tests brought back, as far as possible.

The version from before is still available here: https://github.com/AlwinEsch/kodi/tree/fix-estuary-strings_OLD_NO_REBASE
